<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Шаг-1.-Откройте-файл-с-данными-и-изучите-общую-информацию" data-toc-modified-id="Шаг-1.-Откройте-файл-с-данными-и-изучите-общую-информацию-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Шаг 1. Откройте файл с данными и изучите общую информацию</a></span></li><li><span><a href="#Шаг-2.-Предобработка-данных" data-toc-modified-id="Шаг-2.-Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Шаг 2. Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Обработка-пропусков" data-toc-modified-id="Обработка-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Обработка пропусков</a></span></li><li><span><a href="#Замена-типа-данных" data-toc-modified-id="Замена-типа-данных-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Замена типа данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Лемматизация</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Шаг-3.-Ответьте-на-вопросы" data-toc-modified-id="Шаг-3.-Ответьте-на-вопросы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Шаг 3. Ответьте на вопросы</a></span></li><li><span><a href="#Шаг-4.-Общий-вывод" data-toc-modified-id="Шаг-4.-Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Шаг 4. Общий вывод</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Исследование надёжности заёмщиков

**Цель проекта:** необходимо проанализировать влияние семейного положения и количества детей клиента на факт погашения кредита в срок. Результаты исследования будут учтены при построении модели **кредитного скоринга**.

**Исходные данные:** статистика о платежноспособности клиентов.

**Описание данных:**

* children — количество детей в семье
* days_employed — общий трудовой стаж в днях
* dob_years — возраст клиента в годах
* education — уровень образования клиента
* education_id — идентификатор уровня образования
* family_status — семейное положение
* family_status_id — идентификатор семейного положения
* gender — пол клиента
* income_type — тип занятости
* debt — имел ли задолженность по возврату кредитов
* total_income — ежемесячный доход
* purpose — цель получения кредита

## Шаг 1. Откройте файл с данными и изучите общую информацию

In [1]:
import pandas as pd
df = pd.read_csv("/datasets/data.csv")
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Получим дополнительную информацию о данных с использованием метода describe(). Данный метод позволяет оценить характерные значения и разброс для столбцов, содержащих числовые данные.

In [2]:
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

**Выводы:**

1. В файле содержится таблица с размерами 21525х12.
2. Названия столбцов желательно откорректировать, сделав их более информативными.
3. Пропущенные значения типа Nan содержатся в столбцах "days_employed" и "total_income".
4. В столбце "days_employed" выявлены явные аномалии: содержатся отрицательные значения, среднее арифметическое даже с учетом наличия отрицательных значений составляет 63046 дней, что существенно превышает возможный срок жизни человека (при 100 годах жизни, срок жизни в днях составляет, без учета високосных лет, 365х100 = 36500 дней).
5. В столбце "education" используется одновременно и верхний и нижний регистр.
6. В столбце "purpose" одни и те же цели описаны по разному ("операции с жильем", "покупка жилья" и др.).
7. В столбце "сhildren" имеется как минимум одно отрицательное значение. Кроме того обращает на себя внимание очень большое максимальное значение детей - 20.
8. В столбце "dob_years" имеется как минимум одно нулевое значение - очевидно, что возраст заемщика не может быть равным нулю. 

## Шаг 2. Предобработка данных

### Обработка пропусков

Переименуем названия отдельных столбцов исходной таблицы:

In [3]:
df.rename(columns={"children": "children_number", "dob_years": "age_years", "education": "education_level",
    "education_id": "education_level_id", "purpose": "loan_purpose"}, inplace=True)

Выведем новые названия столбцов таблицы:

In [4]:
print("Новые названия столбцов:")
print(*df.columns, sep="\n")

Новые названия столбцов:
children_number
days_employed
age_years
education_level
education_level_id
family_status
family_status_id
gender
income_type
debt
total_income
loan_purpose


С учетом выявленных аномалий в столбце "children_number" рассмотрим его подробнее:

In [5]:
df["children_number"].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children_number, dtype: int64

Как видно из полученных результатов, в исходной базе данных имеется 47 клиентов с "-1" ребенком. Вероятно, фактическое значение количества детей здесь равняется 1. Выполним соответствующую замену:

In [6]:
df["children_number"] = df["children_number"].replace(-1, 1)

Кроме того в исходной базе данных имеется 76 клиентов с 20 детьми. Указанные значения выглядят недостоверными как в силу большой величины, так и в силу аномального распределения: между 5 и 20 - все значения пропущены. Вероятно, здесь имеется ошибка. В пользу этого говорит также то, что среди 76 родителей с 20 детьми 22 человека моложе 35 лет:

In [7]:
print("Количество людей с 20 детьми моложе 35 лет: {}".format(
    df[(df["children_number"] == 20) & (df["age_years"] < 35)]["age_years"].count()))

Количество людей с 20 детьми моложе 35 лет: 22


Достоверно установить фактическое количество детей для рассматриваемой группы клиентов с 20 детьми не представляется возможным. Условно выполним замену 20 на 2, так как возможно ошибка состоит в лишнем нуле. Данный подход здесь допустим, так как 2 не сильно отклоняется от характерного значения (среднее арифметическое равно 0,54), а также вследствие того, что рассматриваемая выборка из 76 клиентов относительно мала по отношению ко всей выборки из 21525 клиентов.

In [8]:
df["children_number"] = df["children_number"].replace(20, 2)

Подсчитаем количество значений типа Nan в столбце "days_employed":

In [9]:
print("Количество значений типа Nan в столбце 'days_employed': {}".format(df["days_employed"].isnull().sum()))

Количество значений типа Nan в столбце 'days_employed': 2174


Для того чтобы понять, каким образом оптимально заполнить пропуски в столбце "days_employed", рассмотрим его подробнее. При этом для анализа воспользуемся мерами центральной тенденции, а именно медианной с учетом ее большей устойчивости к выбросам. С учетом наличия отрицательных значений отдельно выполним расчет для положительных и отрицательных значений рассматриваемого столбца. Дополнительно также подсчитаем количество отрицательных значений.

In [10]:
print("Количество отрицательных значений по столбцу 'days_employed': {}".format(df[df["days_employed"] < 0].shape[0]))
print("Медиана по столбцу 'days_employed' для всех значений: {:.0f}".format(df["days_employed"].median()))
print("Медиана по столбцу 'days_employed' только для неотрицательных значений: {:.0f}".format(df[df["days_employed"] >= 0]["days_employed"].median()))
print("Медиана по столбцу 'days_employed' только для отрицательных значений: {:.0f}".format(df[df["days_employed"] < 0]["days_employed"].median()))

Количество отрицательных значений по столбцу 'days_employed': 15906
Медиана по столбцу 'days_employed' для всех значений: -1203
Медиана по столбцу 'days_employed' только для неотрицательных значений: 365213
Медиана по столбцу 'days_employed' только для отрицательных значений: -1630


Медиана для неотрицательных значений равняется 365213, что примерно соответствует непрерывной работе в течение 100 лет. Очевидно, что данные в столбце совершенно недостоверны. В качестве возможной причины можно предположить ошибку в единице измерений (дни вместо часов). Принимая во внимание что данные, представленные в столбце 'days_employed', не используются напрямую в настоящем исследовании, пропуски в рассматриваемом столбце условно примем равными нулю.

Замена пропусков в столбце 'days_employed' нулевыми значениями: 

In [11]:
df["days_employed"] = df["days_employed"].fillna(0)

Подсчитаем количество значений типа Nan в столбце "total_income":

In [12]:
print("Количество значений типа Nan в столбце 'total_income': {}".format(df["total_income"].isnull().sum()))

Количество значений типа Nan в столбце 'total_income': 2174


В таблице имеется существенное количество значений типа Nan в столбце "total_income". Причина их появления неизвестна. Можно предположить какую-то техническую ошибку при формировании таблицы. В пользу этой версии говорит то, что в столбце "days_employed" содержится такое же количество значений типа Nan. 


Насчет того, какие фактические значения соответствуют пропускам в столбце "total_income", можно выдвинуть две гипотезы. Первая: данные значения соответствуют нулевому доходу - в этом случае пропуски должны быть заполнены нулями. Вторая гипотеза состоит в том, что фактические доходы в пропущенных строках не равны нулю. В этом случае заполнять пропущенные значения необходимо характерными значениями. Для того, чтобы понять какая из гипотез более обоснованная, подсчитаем количество различных типов занятости по столбцу "income_type" для строк со значениями Nan.

In [13]:
print(df[df["total_income"].isnull()]["income_type"].value_counts())

сотрудник          1105
компаньон           508
пенсионер           413
госслужащий         147
предприниматель       1
Name: income_type, dtype: int64


Из полученных результатов можно сделать следующий вывод: для рассматриваемых строк со значениями Nan основной тип дохода это "сотрудник", "компаньон" и "пенсионер", то есть люди, имеющие доход. Безработные отсутствуют полностью. Следовательно, фактические доходы в пропущенных строках не равну нулю и их необходимо заполнить характерными значениями. При расчете характерных значений будем учитывать тип дохода, в силу его очевидного влияния на полный доход.

Сформируем таблицу "income_medians" с медианными значениями дохода в зависимости от типа дохода:

In [14]:
income_medians = df.groupby("income_type")["total_income"].median()
print(income_medians)

income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        150447.935283
компаньон          172357.950966
пенсионер          118514.486412
предприниматель    499163.144947
сотрудник          142594.396847
студент             98201.625314
Name: total_income, dtype: float64


Запустим цикл, заменяющей значения типа Nan медианными значениями, взятыми из сформированной нами таблицы "income_medians":

In [15]:
for income_type in income_medians.index:
    mask = df.loc[:, "income_type"] == income_type
    df.loc[mask, "total_income"] = df.loc[mask, "total_income"].fillna(income_medians[income_type])

В столбце "age_years" при изучении общей информации была выявлена аномалия - как минимум одно нулевое значение. Подсчитаем сколько в базе данных всего клиентов с нулевым возрастом:

In [16]:
print("Количество клиентов в базе данных c нулевым возрастом: {}".format(df[df["age_years"] == 0]["age_years"].count()))

Количество клиентов в базе данных c нулевым возрастом: 101


С учетом того, что согласно законам РФ заемщиком может быть лицо не младше 18 лет, дополнительно подсчитаем количество клиентов в базе данных младше 18 лет:

In [17]:
print("Количество клиентов в базе данных моложе 18 лет: {}".format(df[df["age_years"] < 18]["age_years"].count()))

Количество клиентов в базе данных моложе 18 лет: 101


Выводы по столбцу "age_years":
1. Количество клиентов младше 18 лет точно совпадает с количеством клиентов с нулевым возрастом, следовательно все клиенты в базе данных младше 18 лет имеют нулевой возраст.
2. Нулевой возраст клиентов является явной ошибкой. Возможная причина данной ошибки - заполнение отсутствующей информации о возрасте клиентов нулевыми значениями. Для исправления данной ошибки нулевые значения возраста необходимо заменить характерными значениями, то есть средними арифметическими.

При расчете характерных значений возраста (для повышения точности расчетов) учтем возможную зависимость возраста от семейного статуса. Для этого рассчитаем средние возраста в зависимости от семейного статуса и запишем их в таблицу "age_means":

In [18]:
age_means = df.groupby("family_status")["age_years"].median()
print(age_means)

family_status
Не женат / не замужем    36
в разводе                46
вдовец / вдова           58
гражданский брак         41
женат / замужем          43
Name: age_years, dtype: int64


Из полученных результатов очевидно, что гипотеза о зависимости возраста от семейного статуса подтвердилась. Учтем это и запустим цикл, заменяющей нулевые значения возраста средними значениями, взятыми из сформированной нами таблицы "age_means":

In [19]:
for family_status in age_means.index:
    mask = (df.loc[:, "family_status"] == family_status) & (df.loc[:, "age_years"] == 0)
    df.loc[mask, "age_years"] = age_means[family_status]

Проконтролируем результаты обработки пропусков:

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children_number       21525 non-null int64
days_employed         21525 non-null float64
age_years             21525 non-null int64
education_level       21525 non-null object
education_level_id    21525 non-null int64
family_status         21525 non-null object
family_status_id      21525 non-null int64
gender                21525 non-null object
income_type           21525 non-null object
debt                  21525 non-null int64
total_income          21525 non-null float64
loan_purpose          21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Выводы:**
1. Скорректированы названия столбцов.
2. В столбце "children_number" скорректированы аномальные значения.
3. Заполнены пропуски в столбцах: "days_employed", "total_income" и "age_years".

### Замена типа данных

В столбцах "days_employed" и "total_income" заменяем вещественный тип данных на целочисленный, применяя метод df.astype(). Применение указанного метода обусловлено его универсальностью.

In [21]:
df["days_employed"] = df["days_employed"].astype("int")
df["total_income"] = df["total_income"].astype("int")

Проконтролируем полученный результат замены типа данных:

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children_number       21525 non-null int64
days_employed         21525 non-null int64
age_years             21525 non-null int64
education_level       21525 non-null object
education_level_id    21525 non-null int64
family_status         21525 non-null object
family_status_id      21525 non-null int64
gender                21525 non-null object
income_type           21525 non-null object
debt                  21525 non-null int64
total_income          21525 non-null int64
loan_purpose          21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


**Вывод**: выполнена замена вещественного типа данных в столбцах "days_employed" и "total_income" на целочисленный.

### Обработка дубликатов

Подсчитаем количество дубликатов в таблице:

In [23]:
print("Количество дубликатов в таблице: {}".format(df.duplicated().sum()))

Количество дубликатов в таблице: 55


Метод df.duplicated() выявляет только полные дубликаты. Однако могут быть дубликаты, отличающиеся используемым регистром. Таким образом, возможно общее количество дубликатов превышает 55. Учитывая это, приведем строковые данные в таблице к нижнему регистру. Выполним это запустив цикл по всем колонкам таблицы, применяя конструкцию "try-except", для того, чтобы пропустить столбцы, содержащие данные числового типа:

In [24]:
for col in df.columns:
    try:
        df[col] = df[col].str.lower()
    except AttributeError:
        pass

Подсчитаем еще раз количество дубликатов в таблице:

In [25]:
print("Количество дубликатов в таблице: {}".format(df.duplicated().sum()))

Количество дубликатов в таблице: 72


Таким образом, после приведения строковых данных к нижнему регистру, количество дубликатов выросло. Следовательно, одна из причин появления дубликатов - заполнение одних и тех же текстовых данных разными регистрами. При слиянии данных из разных источников в одну таблицу одни и те же данные множатся за счет разности регистров. Тем не менее наличие полных 55 дубликатов остается не объяснимым и связано, вероятно, с какой-то технической ошибкой при создании базы данных.

Удалим выявленные дубликаты и проверим полученный результат:

In [26]:
df = df.drop_duplicates()
print("В таблице осталось {} дубликатов".format(df.duplicated().sum()))

В таблице осталось 0 дубликатов


**Вывод**: дубликаты выявлены и обработаны.

### Лемматизация

Загрузим библиотеку pymystem3 с функцией лемматизации на русском языке, а также специльный контейнер Counter (для подсчета числа лемматизированных слов):

In [27]:
from pymystem3 import Mystem
m = Mystem()
from collections import Counter

Объединим столбец с данными о целях займа ("loan_purpose") в единый текст и выполним лемматизацию данного текста:

In [28]:
text = " ".join(df["loan_purpose"])
lemmas = m.lemmatize(text)

Подсчитаем количество упоминаний различных лемматизированных слов в столбце "loan_purpose":

In [29]:
print(Counter(lemmas))

Counter({' ': 55020, 'недвижимость': 6351, 'покупка': 5897, 'жилье': 4460, 'автомобиль': 4306, 'образование': 4013, 'с': 2918, 'операция': 2604, 'свадьба': 2323, 'свой': 2230, 'на': 2221, 'строительство': 1878, 'высокий': 1374, 'получение': 1314, 'коммерческий': 1311, 'для': 1289, 'жилой': 1230, 'сделка': 941, 'дополнительный': 906, 'заниматься': 904, 'подержать': 853, 'проведение': 767, 'сыграть': 765, 'сдача': 651, 'семья': 638, 'собственный': 635, 'со': 627, 'ремонт': 607, 'приобретение': 461, 'профильный': 436, 'подержанный': 111, '\n': 1})


**Выводы**: 
1. Выполнена лемматизация для столбца "loan_purpose".
2. Выполнен подсчет лемматизированных слов для столбца "loan_purpose".

### Категоризация данных

В исследовании рассматривается всего 4 фактора, влияющих на возврат кредита: количество детей, семейное положение, доход и цели кредита. Первые два фактора не нуждаются в дополнительной категоризации, так как уже категоризированы рациональным образом и на относительно небольшое количество категорий. В категоризации, таким образом, нуждаются два оставшихся фактора: доход и цели кредита.

Категоризацию по доходам проведем следующим образом. Разделим все доходы на три категории: низкие, средние и высокие. Для проведения границ между данными категориями воспользуемся квантилями. Примем следующее. Низкий доход - доход меньше квантиля уровня 0,25. Средний доход - доход больше или равно квантиля уровня 0,25, но меньше или равно квантиля уровня 0,75. Высокий доход - доход больше квантиля уровня 0,75. Для определения квантилей воспользуемся методом df.describe():

In [30]:
print("Информация о распределении доходов клиентов по величине:")
df["total_income"].describe()

Информация о распределении доходов клиентов по величине:


count    2.145300e+04
mean     1.653206e+05
std      9.818947e+04
min      2.066700e+04
25%      1.076200e+05
50%      1.425940e+05
75%      1.958210e+05
max      2.265604e+06
Name: total_income, dtype: float64

Округлив полученные значения квантилей до десятков тысяч, получим следующие границы для доходов. Низкий доход - доход меньше 100000. Средний доход - доход больше или равно 100000 и меньше или равно 200000. Высокий доход - доход свыше 200000. Напишем программу "total_income_group" для соответствующей категоризации клиентов по доходам:

In [31]:
def total_income_group(total_income):
    if total_income < 110000:
        return "низкий доход"
    if 110000 <= total_income <= 200000:
        return "средний доход"
    else:
        return "высокий доход"

Добавим новый столбец "total_income_group" к таблице, в котором укажем категорию дохода клиента:

In [32]:
df["total_income_group"] = df["total_income"].apply(total_income_group)

Напишем программу "loan_purpose_group" для категоризации целей займа:

In [33]:
def loan_purpose_group(loan_purpose):
    lemmas = m.lemmatize(loan_purpose)
    if "недвижимость" in lemmas or "жилье" in lemmas:
        return "недвижимость"
    if "автомобиль" in lemmas:
        return "автомобиль"
    if "образование" in lemmas:
        return "образование"
    if "свадьба" in lemmas:
        return "свадьба"
    else:
        return "категория не определена"

Воспользуемся программой "loan_purpose_group" для категоризации клиентов по целям кредита. Добавим новый столбец с категориями по назначениям займа к общей таблице

In [34]:
df["loan_purpose_group"] = df["loan_purpose"].apply(loan_purpose_group)

Выведем на экран таблицу, дополненную двумя столбцами:

In [35]:
df.head()

children_number  days_employed  age_years education_level  \
0                1          -8437         42          высшее   
1                1          -4024         36         среднее   
2                0          -5623         33         среднее   
3                3          -4124         32         среднее   
4                0         340266         53         среднее   

   education_level_id     family_status  family_status_id gender income_type  \
0                   0   женат / замужем                 0      f   сотрудник   
1                   1   женат / замужем                 0      f   сотрудник   
2                   1   женат / замужем                 0      m   сотрудник   
3                   1   женат / замужем                 0      m   сотрудник   
4                   1  гражданский брак                 1      f   пенсионер   

   debt  total_income                loan_purpose total_income_group  \
0     0        253875               покупка жилья      высокий доход   
1     0        112080     приобретение автомобиля      средний доход   
2     0        145885               покупка жилья      средний доход   
3     0        267628  дополнительное образование      высокий доход   
4     0        158616             сыграть свадьбу      средний доход   

  loan_purpose_group  
0       недвижимость  
1         автомобиль  
2       недвижимость  
3        образование  
4            свадьба

Проконтролируем результаты выполненной категоризации данных по полному доходу (total_income) и целям займа (loan_purpose):

In [36]:
print("Результы категоризации данных по полному доходу (total_income):")
print(df["total_income_group"].value_counts(), end="\n\n")
print("Результы категоризации данных по целям кредита (loan_purpose):")
print(df["loan_purpose_group"].value_counts())

Результы категоризации данных по полному доходу (total_income):
средний доход    10744
низкий доход      5642
высокий доход     5067
Name: total_income_group, dtype: int64

Результы категоризации данных по целям кредита (loan_purpose):
недвижимость    10811
автомобиль       4306
образование      4013
свадьба          2323
Name: loan_purpose_group, dtype: int64


**Вывод:** выполнена категоризация данных по величине полного дохода и целям займа.

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

Для ответа на поставленный вопрос создадим сводную таблицу, группируя данные по оси "0" по количеству детей ("children_number"), а по оси "1" - по факту возврата или не возврата долга ("debt"). Кроме того добавим к сводной таблице дополнительный столбец ("past_due_loans_share") с указанием доли невозвращенных кредитов по отношению ко всем кредитам в рассматриваемой категории. Отсортируем таблицу по данному столбцу по убыванию. В итоге получим следующее:

In [37]:
children_number_debt_pivot = df.pivot_table(index="children_number", columns="debt", values="age_years", aggfunc="count")
children_number_debt_pivot["past_due_loans_share"] = children_number_debt_pivot[1] / (
        children_number_debt_pivot[0] + children_number_debt_pivot[1])
children_number_debt_pivot = children_number_debt_pivot.fillna(0)
children_number_debt_pivot.sort_values(by="past_due_loans_share", ascending=False)

debt                   0       1  past_due_loans_share
children_number                                       
4                   37.0     4.0              0.097561
2                 1926.0   202.0              0.094925
1                 4410.0   445.0              0.091658
3                  303.0    27.0              0.081818
0                13027.0  1063.0              0.075444
5                    9.0     0.0              0.000000

**Вывод:**

Между количестом детей и возвратом кредита в срок по всей видимости прямой зависимости нет. Однако можно отметить, что наименьшая доля просроченных кредиторов соответствует бездетным клиентом (клиенты с 5 детьми не учитывались здесь при анализе, ввиду их небольшого общего количества - всего 9 человек). Возможно на возврат кредит влияет сам факт наличия/отсутствия детей, при этом количество детей не имеет значений. Для более достоверного анализа, вероятно, необходимо провести дополнительные статистические исследования.

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

Для ответа на поставленный вопрос создадим сводную таблицу, группируя данные по оси "0" по семейному положению ("family_status"), а по оси "1" по факту возврата или не возврата долга ("debt"). Кроме того добавим к сводной таблице дополнительный столбец ("past_due_loans_share") с указанием доли невозвращенных кредитов по отношению ко всем кредитам в рассматриваемой категории. Отсортируем таблицу по данному столбцу по убыванию. В итоге получим следующее:

In [38]:
family_status_debt_pivot = df.pivot_table(index="family_status", columns="debt", values="age_years", aggfunc="count")
family_status_debt_pivot["past_due_loans_share"] = family_status_debt_pivot[1] / (
        family_status_debt_pivot[0] + family_status_debt_pivot[1])
family_status_debt_pivot.sort_values(by="past_due_loans_share", ascending=False)

debt                       0    1  past_due_loans_share
family_status                                          
не женат / не замужем   2536  274              0.097509
гражданский брак        3762  388              0.093494
женат / замужем        11408  931              0.075452
в разводе               1110   85              0.071130
вдовец / вдова           896   63              0.065693

**Вывод**

Между семейным статусом и возвратом кредита в срок, вероятно, имеется зависимость. Наибольшая доля невозвратов кредита в срок соответствует статусу "не женат / не замужем", наименьшая - статусу "вдовец / вдова". Для более достоверного анализа, вероятно, необходимо провести дополнительные статистические исследования.

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Для ответа на поставленный вопрос создадим сводную таблицу, группируя данные по оси "0" по уровню дохода ("total_income_group"), а по оси "1" по факту возврата или не возврата долга ("debt"). Кроме того добавим к сводной таблице дополнительный столбец ("past_due_loans_share") с указанием доли невозвращенных кредитов по отношению ко всем кредитам в рассматриваемой категории. Отсортируем таблицу по данному столбцу по убыванию. В итоге получим следующее:

In [39]:
total_income_group_debt_pivot = df.pivot_table(index="total_income_group", columns="debt", values="age_years", aggfunc="count")
total_income_group_debt_pivot["past_due_loans_share"] = total_income_group_debt_pivot[1] / (
        total_income_group_debt_pivot[0] + total_income_group_debt_pivot[1])
total_income_group_debt_pivot.sort_values(by="past_due_loans_share", ascending=False)

debt                   0    1  past_due_loans_share
total_income_group                                 
средний доход       9816  928              0.086374
низкий доход        5187  455              0.080645
высокий доход       4709  358              0.070653

**Вывод**

Между уровнем дохода и возвратом кредита в срок, вероятно, прямой зависимости нет. Наименьшая доля невозвратов кредита в срок соответствует высокому уровню дохода, что логично. Однако вместо ожидаемого результата, что наибольшая доля невозвратов соответствует низкому уровню дохода, мы имеем, что эта доля соответствует среднему доходу. Для более достоверного анализа, вероятно, необходимо провести дополнительные статистические исследования.

- Как разные цели кредита влияют на его возврат в срок?

Для ответа на поставленный вопрос создадим сводную таблицу, группируя данные по оси "0" по уровню дохода ("loan_purpose_group"), а по оси "1" по факту возврата или не возврата долга ("debt"). Кроме того добавим к сводной таблице дополнительный столбец ("past_due_loans_share") с указанием доли невозвращенных кредитов по отношению ко всем кредитам в рассматриваемой категории. Отсортируем таблицу по данному столбцу по убыванию. В итоге получим следующее:

In [40]:
loan_purpose_group_debt_pivot = df.pivot_table(index="loan_purpose_group", columns="debt", values="age_years", aggfunc="count")
loan_purpose_group_debt_pivot["past_due_loans_share"] = loan_purpose_group_debt_pivot[1] / (
        loan_purpose_group_debt_pivot[0] + loan_purpose_group_debt_pivot[1])
loan_purpose_group_debt_pivot.sort_values(by="past_due_loans_share", ascending=False)

debt                    0    1  past_due_loans_share
loan_purpose_group                                  
автомобиль           3903  403              0.093590
образование          3643  370              0.092200
свадьба              2137  186              0.080069
недвижимость        10029  782              0.072334

**Вывод**

Между целями кредита и возвратом его в срок, вероятно, есть зависимость. Наибольшая доля невозвратов кредита в срок соответствует цели "автомобиль", наименьшая - цели "автомобиль". Для более достоверного анализа, вероятно, необходимо провести дополнительные статистические исследования.

## Шаг 4. Общий вывод

Выполнена предобработка полученных исходных данных. Проанализирована зависимость возврата кредита в срок от четырех факторов: количества детей, семейного статуса, уровня дохода и цели кредита. Следует отметить, что для более достоверного анализа, вероятно, необходимо провести дополнительные статистические исследования.

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.